### ISSUE TITLE
Warn about invalid preprocessor expressions by default?

### ISSUE THREAD

**Issue by ojwb** [swig/swig#1465]
It seems SWIG quietly accepts `#if define XXX` (where `define` is an easy typo to make for `defined`).

Testing with current git master (3.0.12 behaves the same way):

```
$ cat test.i
#if define SWIGJAVA
#error bang
#endif
$ ./preinst-swig -module foo -java test.i
$ ./preinst-swig -module foo -python test.i
$
```

I would expect that an invalid expression in `#if` would give an error.

GCC gives an error:

```
$ g++ -xc++ -E test.i
# 1 "test.i"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "test.i"
test.i:1:12: error: missing binary operator before token "SWIGJAVA"
 #if define SWIGJAVA
            ^~~~~~~~
```

**Comment by ojwb** [comment]
Still present in git master.

**Comment by ojwb** [comment]
With `-Wextra` there are two warnings:

```
test.i:1: Warning 202: Could not evaluate expression 'define SWIGJAVA'
test.i:1: Warning 202: Error: 'Syntax error: expected operator'
```

The second is rather misleading as this isn't actually handled as an error - files are generated and exit status is zero.

This is a bit more subtle that I thought actually - e.g. `#if define` is OK with GCC because `define` is just an identifier here.  And while it's perhaps the easiest typo to make for `defined`, why treat `define` differently to `#if defines SWIGJAVA` etc?

I'll at least fix the bogus "Error:" in that warning.

**Comment by ojwb** [comment]
> I'll at least fix the bogus "Error:" in that warning.

Fixed by 74d12d8c4c15519e06c672ab48e72b24acd741f0.

Thinking about the rest, it seems to me the underlying problem here is that SWIG doesn't actually report errors in preprocessor expressions without `-Wextra` or `-Wall`[*].

Ignoring errors in preprocessor expressions by default is presumably geared towards not moaning about invalid expressions in conditionals in header files being wrapped that result from things not being defined when running under SWIG.  However it means we fail to flag errors in preprocessor expressions in SWIG interface files, which seems a problem.

We could perhaps come up with some sort of interface vs header distinction, but really I'm a dubious that such problems are really seen in the wild, especially since undefined macros evaluate as zero so e.g. `#if X` and `#if X > 1` are both OK without `X` being defined.

I can construct examples which fail - e.g. this (where `OP` normally gets defined to `==` or `!=` but isn't defined under SWIG):

```
#if X OP Y
```

However I wonder if it isn't better to actually flag those up and have the user decide how to resolve (suppress the warning, define `OP` in the interface file before parsing the header, or get SWIG to parse some other header to define `OP` suitably first).  Currently we arbitrarily treat this condition as false, which might be good but might be really unhelpful.

---

[*] Aside: it's confusing for anyone familiar with the options of the same name in GCC, clang, etc that in SWIG `-Wall` seems to be a superset of `-Wextra`, while in compilers `-Wextra` enables additional warnings that `-Wall` doesn't - that seems too hard to address at this point though, and the naming in compilers is kind of confusing in itself; clang adds `-Weverything` into the mix, which literally turns on every single implemented warning and is not useful in practice on any real codebase.

**Comment by ojwb** [comment]
Enabling this warning reveals a bug in our own testsuite:

```
diff --git a/Examples/test-suite/preproc_defined.i b/Examples/test-suite/preproc_defined.i
index e1958515..5ebf0a09 100644
--- a/Examples/test-suite/preproc_defined.i
+++ b/Examples/test-suite/preproc_defined.i
@@ -86,7 +86,7 @@ struct Defined {
 void defined_not(TYPE);
 #endif
 
-#if !( defined(AAA) \
+#if !( defined(AAA) &&\
  defined(BBB) \\
 && defined(CCC) )
 void bumpf_not(TYPE);
```

I've opened a PR.
